### DISCUSSION TITLE
swupdate/capsule update, how to mark-good?

### DISCUSSION THREAD

**Discussion by AndersGnistrup** [OE4T/meta-tegra#1654] [Q&A] [answered]
I have a working system based on orin/nano, meta-tegra and swupdate
It works just fine and I can update the system via swupdate -i <image>
I have read this guide:
https://docs.nvidia.com/jetson/archives/r36.2/DeveloperGuide/SD/Bootloader/UpdateAndRedundancy.html#manually-sync-bootloader-a-b-slots-with-the-capsule-update
From what I can read the process is: 
- copy capsule and notify the bootloader setting bit2 of the OsIndications UEFI variable.
- reboot and check the status the capsule update using nvbootctrl

But this only indicates if the update was OK, not if the SW in the packages is OK.
The rauc has the operation mark-good to indicate that SW is good to go after the update.

How is the "mark good" handled in swupdate?

**Comment by AndersGnistrup** [comment]
Ok, after some further reading it seems to be related to a swupdate progress and a swupdate service.
I will post my findings.

**Comment by AndersGnistrup** [comment]
I was wrong, swupdate-progress has nothing to do with mark-good, only the current state/progress of the file transfer/copy.

I am using  origin/scarthgap-l4t-r35.x on meta-tegra
There is a script nv_update_verifier that calls nvbootctrl verify but this returns:
_variable BootChainFwStatus is not found._
And I gees the reason is that the uefi bootloader does not create this variable in the uefi variables.
So, I think I need to update my distro a clean scarthgap branch,

overall conclusion: mark good does not work in jetson 35.x, you need 36.3
Is this correct?

I have used the source for debug and these pages:
https://docs.nvidia.com/jetson/archives/r36.3/DeveloperGuide/SD/RootFileSystem.html#rootfs-selection
https://github.com/OE4T/meta-tegra/pull/1428
https://docs.nvidia.com/jetson/archives/l4t-archived/l4t-3275/index.html#page/Tegra%20Linux%20Driver%20Package%20Development%20Guide/rootfs_custom.html#wwpID0E5HA

**Comment by dwalkes** [comment, marked as answer]
Not sure if you found it already but there is some discussion about `BootChainFwStatus` for marking good status in https://github.com/mendersoftware/meta-mender-community/pull/394 and https://github.com/OE4T/meta-mender-community/pull/19.  I'm not completely sure where this ended up with the current mender implementation but this might be something to look at.  Ideally we'd upstream something all update frameworks could use.

This would be something great to add to the reference swupdate demo at https://github.com/OE4T/tegra-demo-distro?tab=readme-ov-file#update-image-demo.  There's a Jetpack 4 swupdate implementation in [Matt's test distro](https://github.com/madisongh/tegra-test-distro/blob/6b5ad1deff310d9eb0ec0b41e1e87c1bb0301c8b/layers/meta-testdistro/recipes-support/swupdate/swupdate-machine-config/update_bootinfo-mods.conf#L2C15-L2C38) which used [https://github.com/OE4T/tegra-boot-tools/blob/059ddc8187ae0e9f7ba0acdb672bdabaa7559434/doc/tegra-bootinfo.md](tegra-bootinfo) for this.  You can find more details at about that original implementation from Matt which I used for the tegra-demo-distro implementation at https://github.com/madisongh/tegra-test-distro/wiki/swupdate-integration-notes

**Reply by AndersGnistrup** [reply]
Thanks for the reply.
I will  try to make a solution that could be used in the swupdate demo. But that will take some time. At lease I have a known path.

**Comment by AndersGnistrup** [comment]
Now for an update.
First, nvbootctrl verify does work, but emits a warning about the BootChainFwStatus is missing. I am not sure but this seems to be a left over since the efi var is gone in the kernel in jetson 36.3
If I delete/disable the service nv-update_verifyer.service and boots three times after a swupdate, bootloader switches back back to A which is expected.

```nvbootctrl -t rootfs dump-slots-info
Current rootfs slot: A
Active rootfs slot: A
num_slots: 2
slot: 0,             retry_count: 3,             status: normal
slot: 1,             retry_count: 0,             status: unbootable
```
So stuff actually seems to work just fine.
The problem is that the rootfs slot B is marked unbootable so new firmware will be discarded by the bootloader. These are the flags.
```
cat /sys/firmware/efi/efivars/RootfsStatusSlotB-781e084c-a330-417c-b678-38e696380cb9 | hexdump -C
00000000  07 00 00 ff 00 00 00 00                           |........|
00000008
```
There are several ways to clear the flag before a swupdate, see https://forums.developer.nvidia.com/t/rootfs-a-b-redundancy-fail-over-mechanism-in-jetpack5-1/247290

And now I am a bit confused.
It seems that the BootChainFwStatus is obsolete/replaced with something else. Is this correct?
Most, if not all. seems to be covered by the efi vars present in 36.3.

This is right after a swupdate.
```
# nvbootctrl -t rootfs dump-slots-info
Current rootfs slot: B
Active rootfs slot: B
num_slots: 2
slot: 0,             retry_count: 3,             status: normal
slot: 1,             retry_count: 2,             status: normal
# nvbootctrl verify
Info: variable BootChainFwStatus is not found.
# nvbootctrl -t rootfs dump-slots-info
Current rootfs slot: B
Active rootfs slot: B
num_slots: 2
slot: 0,             retry_count: 3,             status: normal
slot: 1,             retry_count: 3,             status: normal
```
